doc/source/development/rfc/rfc97_feature_and_fielddefn_sealing.rst

```diff
@@ -0,0 +1,310 @@
+.. _rfc-97:
+
+===================================================================
+RFC 97: OGRFeatureDefn, OGRFieldDefn and OGRGeomFieldDefn "sealing"
+===================================================================
+
+============== =============================================
+Author:        Even Rouault
+Contact:       even.rouault @ spatialys.com
+Started:       2023-Nov-16
+Status:        In development, not implemented
+Target:        GDAL 3.9
+============== =============================================
+
+Summary
+-------
+
+This RFC aims at avoiding common misuse of the setter methods of the
+OGRFeatureDefn, OGRFieldDefn and OGRGeomFieldDefn classes. Indeed, the setter
+methods of those classes should not be used directly by user code (that is
+non driver implementations), on instances that are owned by a OGRLayer. It is
+quite frequent for users (even seasoned ones) to neglect that constraint. Hence
+this RFC introduces an optional "sealing" capability that drivers can enable to
+avoid users modifying instances that they should not.
+
+Motivation
+----------
+
+Let's take an example to clarify. We want to prevent a user from doing the
+following:
+
+.. code-block:: c++
+
+    poLayer->GetLayerDefn()->GetFieldDefn(iFieldIdx)->SetName("new_name");
```

Comment:
> What would be the consequences of removing `OGRFieldDefn OGRFeatureDefn::GetFieldDefn(int iField)` and leaving only `const OGRFieldDefn *OGRFeatureDefn::GetFieldDefn(int iField) const` ?

Good remark. Just after writing the RFC, I suddenly thought about that, but indeed the issue is that here is no way to enforce const correctness through the C API / SWIG.
Leaving only a const accessor would also likely require lots of changes because driver code might call GetLayerDefn() with the intent of modifying it.

Reply:
I've just added a paragraph in the RFC about that